src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java

```diff
@@ -293,36 +294,42 @@ public SCM build(@NonNull SCMHead head, @CheckForNull SCMRevision revision) {
         }
         return result;
     }
-    
+
     /**
      * Returns true if the branchName isn't matched by includes or is matched by excludes.
-     * 
+     *
      * @param branchName
      * @return true if branchName is excluded or is not included
      */
     protected boolean isExcluded (String branchName){
       return !Pattern.matches(getPattern(getIncludes()), branchName) || (Pattern.matches(getPattern(getExcludes()), branchName));
     }
-    
+
     /**
-     * Returns the pattern corresponding to the branches containing wildcards. 
-     * 
+     * Returns the pattern corresponding to the branches containing wildcards.
+     *
      * @param branchName
      * @return pattern corresponding to the branches containing wildcards
      */
     private String getPattern(String branches){
       StringBuilder quotedBranches = new StringBuilder();
-      for (String wildcard : branches.split(" ")){
+      for (String term : branches.split(" ")){
         StringBuilder quotedBranch = new StringBuilder();
-        for(String branch : wildcard.split("\\*")){
-          if (wildcard.startsWith("*") || quotedBranches.length()>0) {
-            quotedBranch.append(".*");
-          }
-          quotedBranch.append(Pattern.quote(branch));
-        }
-        if (wildcard.endsWith("*")){
-          quotedBranch.append(".*");
+
+        if (term.startsWith("^")) { // Regex mode, ^ is not a valid character in git names
```

Comment:
I think the fundamental problem is that you want to introduce an incompatible change (make "master" mean something different in this context than it does in other contexts in the plugin) and I'm unwilling to accept an incompatible change.

In the bug report, you said:

> For example, including 'master' matches both 'master' & 'foo/master'.

That is the behavior of the existing plugin if I enter "master" as the branch to be built.  I realize that is implicit wildcard behavior, but with over 60000 installations, I'm not willing to change the behavior of the plugin in its handling of that case.

I don't especially like the behavior personally, but with that many installations, there will be more people harmed by the change than will be helped by the change.

I recommend closing this pull request rather than changing behavior